src/Adapter/AbstractObjectModelRepository.php

```diff
@@ -73,28 +79,82 @@ protected function assertObjectModelExists(int $id, string $objectTableName, str
      */
     protected function getObjectModel(int $id, string $objectModelClass, string $exceptionClass): ObjectModel
     {
-        try {
-            $objectModel = new $objectModelClass($id);
+        return $this->fetchObjectModel($id, $objectModelClass, $exceptionClass, null);
+    }
 
-            if ((int) $objectModel->id !== $id) {
-                throw new $exceptionClass(sprintf('%s #%d was not found', $objectModelClass, $id));
-            }
-        } catch (PrestaShopException $e) {
-            throw new CoreException(
-                sprintf(
-                    'Error occurred when trying to get %s #%d [%s]',
-                    $objectModelClass,
-                    $id,
-                    $e->getMessage()
-                ),
-                0,
-                $e
-            );
-        }
+    /**
+     * @param int $id
+     * @param string $objectModelClass
+     * @param string $exceptionClass
+     *
+     * @return ObjectModel
+     *
+     * @throws CoreException
+     */
+    protected function getObjectModelForShop(int $id, string $objectModelClass, string $exceptionClass, ShopId $shopId): ObjectModel
+    {
+        $objectModel = $this->fetchObjectModel($id, $objectModelClass, $exceptionClass, $shopId->getValue());
+
+        // The object is fetched before checking the association, so that the NotFoundException has the priority over the NoAssociationException
+        $this->checkShopAssociation($id, $objectModelClass, $shopId);
+
+        // Force id_shop_list right away so that DB modification use the appropriate shop and not the one from context
+        $objectModel->id_shop_list = [$shopId->getValue()];
 
         return $objectModel;
     }
 
+    /**
+     * @param int $id
+     * @param string $objectModelClassName
+     * @param ShopId $shopId
+     *
+     * @return bool
+     *
+     * @throws ShopAssociationNotFound
+     */
+    protected function hasShopAssociation(int $id, string $objectModelClassName, ShopId $shopId): bool
```

Comment:
ObjectModel::isAssociatedToShop?

Reply:
I didn't know of that one either, I will refacto to use it Thanks

Reply:
So after checking I can't use `ObjectModel::isAssociatedToShop` in this situation, it only works on an instanciated object, like `$product->isAssociatedToShop($shopId)`

But here it is not the exact same use case, I need to check the existence based on IDs and object type alone to avoid relying on a instance of ObjectModel, that's why I think it's better to keep it that way here